src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.proxy.callback;
+
+import io.r2dbc.spi.Batch;
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.Result;
+import io.r2dbc.spi.Statement;
+import io.r2dbc.spi.Wrapped;
+
+/**
+ * @author Tadaya Tsuyukubo
+ */
+public class ProxyUtils {
+
+    private ProxyUtils() {
+    }
+
+    @SuppressWarnings("unchecked")
+    public static Connection getOriginalConnection(Connection connection) {
+        if (connection instanceof Wrapped) {
+            return ((Wrapped<Connection>) connection).unwrap();
+        }
+        return connection;
+    }
+
+    public static Connection getOriginalConnection(Batch<?> batch) {
+        if (batch instanceof ConnectionHolder) {
+            return ((ConnectionHolder) batch).getOriginalConnection();
+        }
+        return null;
+    }
+
+    public static Connection getOriginalConnection(Statement<?> statement) {
+        if (statement instanceof ConnectionHolder) {
+            return ((ConnectionHolder) statement).getOriginalConnection();
+        }
+        return null;
+    }
+
+    public static Connection getOriginalConnection(Result result) {
+        if (result instanceof ConnectionHolder) {
+            return ((ConnectionHolder) result).getOriginalConnection();
+        }
+        return null;
+    }
```

Comment:
Is this `null` approach the right approach? Does wrapping with `Optional` cost too much?

Reply:
Good Idea. I'll convert to Optional

Reply:
Check first. In some scenarios Optional adds a bit of overhead. They actually backed off doing it in Spring Data Commons.

Reply:
These method are utility for users and more for corner case use to retrieve original `Connection` object.
Not in the hot path. So, converted to `Optional`.

Reply:
I also think non-hot code paths are fine for `Optional<…>` usage.